pkg/virt-controller/watch/vm_initializer.go

```diff
@@ -0,0 +1,302 @@
+package watch
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/api/core/v1"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/cache"
+
+	kubev1 "kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+type VirtualMachineInitializer struct {
+	vmPresetInformer cache.SharedIndexInformer
+	clientset        kubecli.KubevirtClient
+}
+
+const initializerMarking = "presets.virtualmachines.kubevirt.io"
+
+// FIXME: Both the restClient and clientSet are probably not needed.
+func NewVirtualMachineInitializer(vmPresetInformer cache.SharedIndexInformer, clientset kubecli.KubevirtClient) *VirtualMachineInitializer {
+	vmi := VirtualMachineInitializer{
+		vmPresetInformer: vmPresetInformer,
+		clientset:        clientset,
+	}
+	return &vmi
+}
+
+func (c *VirtualMachineInitializer) initializeVirtualMachine(vm *kubev1.VirtualMachine) error {
+	// All VM's must be marked as initialized or they are held in limbo forever
+	// Collect all errors and defer returning until after the update
+	logger := log.Log
+	errors := []error{}
+
+	logger.Object(vm).Info("Initializing VirtualMachine")
+
+	allPresets, err := listPresets(c.vmPresetInformer, vm.GetNamespace())
+
+	matchingPresets, err := filterPresets(allPresets, vm)
+
+	if err != nil {
+		logger.Object(vm).Reason(err).Errorf("Error while matching presets to VirtualMachine")
+		errors = append(errors, err)
+	}
+
+	if len(matchingPresets) != 0 {
+		err = applyPresets(vm, matchingPresets)
```

Comment:
Hrm. actually this is a gap in the algorithm so far. I believe the correct approach is to apply the presets in the order the labels are listed in the VM's metadata. This will give the user a concise and understandable method of ensuring they have control of what they wanted to specify.

Reply:
We'd need to be certain label ordering is something that is consistent if we take this route. I don't know if we can depend on that.

Reply:
if we go that route (layering multiple presets over) then it really needs to be clear for users how it's done and they need control over it.
I'm not sure we should go this route either, though.